### ISSUE TITLE
Batched Anchor Data Structure

### ISSUE THREAD

**Issue by oed** [ceramicnetwork/CIP#69]
```
cip: <to be assigned>
title: Batched Anchor Data Structure
author: Joel Thorstensson (@oed)
discussions-to: <URL of the Github issue for this CIP (if this is a PR)>
status: Draft
category: Standards
type: Core
created: 2020-11-02
```

## Simple Summary
This CIP describes the data structure of a blockchain anchor that batches multiple document updates together. 

## Abstract
An anchor service accepts document update anchor request from clients and batches them together in a single merkle tree, which is then anchored into a specific blockchain. The anchor service also sorts all of the requests and creates a bloom filter which describes which documents where updated in this batch, which is also included in the blockchain transaction. This allows external observers to more effectively determine if a document which they are interested in was updated.

## Motivation
There are two main motivations for this CIP. Firstly having an agreed upon spec for the structure of the batched anchor merkle tree makes it easier for all consumers of the ceramic protocol to know what to expect when parsing an *anchor record*. Secondly the sorting of updates and the inclusion of a bloom filter allows indexing services to effectively determine if an update was made which is relevant, and find this update in the tree effectively.

## Specification

First lets define the basic IPLD data structures used.
```ipldsch
type MerkleNode [Link]

type TreeMetaData struct {
  numEntries Int
  bloomType String
  bloomFilter {String:Any}
}
```
Here the *MerkleNode* is simply an array of CIDs, which is used as the main building block for the merkle tree. The *TreeMetaData* contains some metadata about the tree, namely the number of entries and a bloom filter. The `bloomType` is used to describe which specific type of bloom filter is being used and is there for upgradeability purposes.

### Tree structure
The structure of the merkle tree is very straight forward. Starting from the bottom each individual update is paired with another update and a *MerkleNode* is created which has the leftmost update in position `0`, and the other update at position `1`. *MerkleNodes* are then created on each level until we reach the root. The root *MerkleNode* has an additional entry at position `2` which simply points to a *TreeMetaData* object. In the figure below we see a simple tree with four leafs.
![merkletree(1)](https://user-images.githubusercontent.com/3909429/97851426-a83bf900-1cf5-11eb-9d9f-e7b568138850.png)
In this example an *IPLD Path* can be used to reach any of the leafs. For example, to reach *Update 3* we would use the path `<root-cid>/1/0`. Similarly we can always get the *TreeMetaData* using the path `<root-cid>/2`.

In the case where the number of leafs are equal to `2^n` it's very simple to construct the tree. However if it's not we take the following approach. **----->TODO----->**

### Leaf sorting
Before creating the tree the leafs (updates) should be sorted. The sorting should be based on a few different properties of the Ceramic document that is being updated:

1. `tags` - sort by the first tag in the *tags* array, then sort by any additional tag in the array
2. `schema` - if multiple documents have the same tag, sort by the *schema*
3. `controllers` - if multiple documents have the same tag and schema, sort by the first *controller*, then sort by any additional *controller*
4. `DocID` - finally sort by the DocID

### Bloom filter
The bloom filter should include the following data for each document that is being updated:

* The first 5 `tags`
* The `schema`
* All DIDs in the `controllers` array
* The DocID

The bloom filter is created using the javascript [bloom-filter](https://github.com/Callidon/bloom-filters) library. Specifically using the *Classic Bloom Filter*. An example for how to create the filter can be observed below.

```js
const { BloomFilter } = require('bloom-filters')
const filterEntries = // ...
const errorRate = 0.04 // 4 % error rate
const filter = BloomFilter.from(items, errorRate)
const exported = filter.saveAsJSON()
```

Now we can create the *TreeMetaData* object, which can be added into ipfs and the CID of which is added to position `2` in the root *MerkleNode*.

```js
{
  numEntries: <number-of-leafs>,
  bloomType: 'npmjs_bloom-filter',
  bloomFilter: exported
}
```

## Rationale
This CIP provides the most efficient merkle tree implementation possible in IPLD. The more interesting aspects of this CIP are however the sorting algorithm and the bloom filter implementation. 

The sorting allows for more efficient lookup in the merkle tree, primarily based on the primary tag of the updated Ceramic document. Further sorting is also provided by the additional properties. If an implementer knows the first tag of a document they can do a binary search though the tree to find the correct update.

The bloom filter allows an observer to efficiently determine if any of the given properties are most likely in the merkle tree. Since the filter itself is stored in a separate IPLD object the filter doesn't add any overhead when verifying a merkle witness for a document update. The bloomfilter implementation that was chosen is currently only implemented in javascript, we failed to find an implementation that is compatible with multiple languages. However, since we introduce a `bloomType` property we can easily determine which implementation was used. This also allows us to change the implementation used in the future.

## Backwards Compatibility
These changes should be fully backwards compatible with the current implementation of Ceramic.

## Implementation
* [Ceramic Anchor Service](https://github.com/ceramicnetwork/ceramic-anchor-service) - Not started yet.

## Security Considerations
The main consideration to take into account when implementing the *Batched Anchor Data Structure* is to have a limit on the number of leafs, i.e. tree depth. If the tree is too large it will become more expensive to verify the merkle witness for any individual update. 
In the bloom filter we limit the `tags` inclusion to the first 5 tags. The reason for this is simply to limit the size of the generated bloom filter. 

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by stbrody** [comment]
@oed, can you elaborate on what exactly your question is about building the merkle tree when there are different numbers of leaf nodes?

Beyond that, I'd like to see a bit more about what exactly the entries in the bloom filter look like/how they are built.  Your code example actually has that part commented out and unspecified.  Are you imagining that the tags, schema, controllers, and docId all get encoded into one entry somehow?  Or does one ceramic document create like 4-10 different entries into the bloom filter?  If the latter, do you plan to add anything into how these entries are encoded to allow distinguishing between entries from the different input types (distinguishing an entry that is a CID based on whether it was a main docID, a schema reference docId, or a tag, for example).

Lastly, I'm a little confused as to how one would actually take advantage of the sortedness of the merkle tree, given that when you're at an intermediate node in the tree, you don't actually have any information encoded in that node about what portion of the sorted keyspace it represents, so you'd have no way to actually decide which branch of the tree to follow to get closer to the set of documents you are looking for, unless I'm missing something?

**Comment by oed** [comment]
> can you elaborate on what exactly your question is about building the merkle tree when there are different numbers of leaf nodes?

Basically if the number of nodes is not a power of 2 then the tree will not be symmetrical. We should however always form the tree in a deterministic way, so which algorithm should we use for that? cc @simonovic86 

> Beyond that, I'd like to see a bit more about what exactly the entries in the bloom filter look like/how they are built.

It's an array of strings, updating the first post. Each document will create multiple entries in the filter.

>  If the latter, do you plan to add anything into how these entries are encoded to allow distinguishing between entries from the different input types

Good point! We need a way to distinguish between DocIDs and Schema-DocIDs at the very least!

> Lastly, I'm a little confused as to how one would actually take advantage of the sortedness of the merkle tree,

Yeah since there are more than one tag this is not optimal I think. Basically if you know that a tag is the first tag of a document then you can do a binary search to effectively find the update. However, if you only know some other info you can't really take much advantage of this. Might be better options here so open to suggestions!

**Comment by simonovic86** [comment]
@oed @stbrody we already cover the case of odd number of leaves. This is the standard way of constructing merkle trees if I'm not mistaken. Here's the test case: [link](https://github.com/ceramicnetwork/ceramic-anchor-service/blob/4615c814ed98bb17cba8902480d8f22bde5397a1/src/merkle/__tests__/merkle-layers.test.ts#L70).

**Comment by stbrody** [comment]
Yeah, the algorithm for building a balanced binary search tree is pretty standard and straightforward (I used to ask it as an interview question :P), I believe the only part of it that could be different implementation by implementation is when you have an even number of nodes whether you take the rightmost node in the left half as the root, or the leftmost node of the right half.  Example code [here](https://gist.github.com/stbrody/bf09a00cfed7223401c66b1032929db5).  As long as we're consistent though it should be fine either way I'd think.

@oed I see you updated the description above to include prepending "tag-" and "schema-" to those bloom filter records - I'd suggest we do the same for the other entries too ("controller-did-" and "DocId-") to future-proof ourselves in case we want to add additional DIDs or DocIds into the bloom filter in the future.  Adding characters to the size of the entries in the bloom filter is free, since the actual entry strings never make it into the filter, just the hashes of them do.

My concern about doing binary search on the tree is about more than having access to the right information of the document you're looking up.  Even if you have all the fields of the document you are looking for, I'm still not sure how you could actually use that to search in the tree.  The problem is that at the non-leaf nodes of the tree, you only have the CIDs of your left and right sub-tree nodes, the tree itself has no information about the contents of the documents covered by each subtree.  To use the sortedness property of the tree, the non-leaf nodes would need to also contain information about which range of the overall sorted keyspace they cover

**Comment by oed** [comment]
Created a PR for this CIP, updated the spec and made this into the discussion thread. See #74 
Note that this introduces the `topic` property in the metadata of documents, which is currently not implemented anywhere. Some work remains to add this in various places.
